### ISSUE TITLE
Preprocessing language+mathematics corpora for pretraining

### ISSUE THREAD

**Issue by dginev** [KWARC/llamapun#59]
A good 2020 use of llamapun would be to use it as a unified preprocessing step for a variety of HTML corpora which also include math syntax by one trick or another. The goal would be to do the legwork on a variety of HTML dialects so that we get clean and maximally denoised data as a plaintext target, with a primary focus on using that textual form for pretraining a neural language model. I will be using this issue as a documentation placeholder for the various targets I have in mind.

**Comment by holtzermann17** [comment]
Maybe add: [proofwiki](https://proofwiki.org/wiki/Main_Page)?

**Comment by dginev** [comment]
Thanks @holtzermann17 , interesting suggestion. I will also "audit" that resource. I've added the link to their latest XML dump in the issue description and downloading now. Doing a gradual step-by-step auditing, preprocessing and packaging into pretrain-friendly versions in the coming weeks.

So please feel free (and anyone else reading here!) to link me to any other large-ish resources with math syntax that feel similar to the list above. And thanks!

**Comment by dginev** [comment]
Here is an unfortunate example of how s2orc deals with mathematical formulas, a fundamental limitation of scraping PDFs. You'll see scripts silently put on the baseline, badly tokenized paragraphs when there is display math, and of course missing markup. A non-starter in my eyes.

```json
{
   "section":"Theorem 2.1. For the solution u to Problem (1.1) with nonnegative weights f , there is a sufficiently smooth function S(x, t) such that u(x, t) < S(x, t), t > 0. (2.3)",
   "text":"The choice of the function S(x, t) depends on the weights and sup x\u2208\u2126 u 0 (x).",
   "cite_spans":[],"ref_spans":[]},
{
   "section":"Theorem 2.1. For the solution u to Problem (1.1) with nonnegative weights f , there is a sufficiently smooth function S(x, t) such that u(x, t) < S(x, t), t > 0. (2.3)",
   "text":"Proof. Let S = p + q with p and q such that",
   "cite_spans":[],"ref_spans":[]},
{
   "section":"Theorem 2.1. For the solution u to Problem (1.1) with nonnegative weights f , there is a sufficiently smooth function S(x, t) such that u(x, t) < S(x, t), t > 0. (2.3)",
   "text":"where r = |x| = x 2 1 + x 2 2 1/2 . The positive constants A, a and \u03ba will be chosen later. Then p and q satisfy",
   "cite_spans":[],"ref_spans":[]},
{
   "section":"Theorem 2.1. For the solution u to Problem (1.1) with nonnegative weights f , there is a sufficiently smooth function S(x, t) such that u(x, t) < S(x, t), t > 0. (2.3)",
   "text":"If a satisfies ",
   "cite_spans":[],"ref_spans":[]},
{
   "section":"Theorem 2.1. For the solution u to Problem (1.1) with nonnegative weights f , there is a sufficiently smooth function S(x, t) such that u(x, t) < S(x, t), t > 0. (2.3)",
   "text":"Then, by the choice of \u03ba, k and a, we see that S = p + q satisfies",
   "cite_spans":[],"ref_spans":[]},
{
   "section":"Theorem 2.1. For the solution u to Problem (1.1) with nonnegative weights f , there is a sufficiently smooth function S(x, t) such that u(x, t) < S(x, t), t > 0. (2.3)",
   "text":"and by the inequality (2.6), we obtain for z \u2208 \u2202\u2126,",
   "cite_spans":[],"ref_spans":[]},
{
   "section":"S(z, t) = p(z, t) + q(z, t) =",
   "text":"A e \u03bat + e \u03bat e \u22122a > Ae \u03bat > f (z, \u00b7) 2 ( p(\u00b7, t) 2 + q(\u00b7, t) 2 ) \u2265 \u2126 f(z, y)(p + q) dy = \u2126 f (z, y",
   "cite_spans":[],"ref_spans":[]},
{
   "section":")S(y, t) dy, 0 < t < T.",
   "text":"Note that above inequalities hold for arbitrary positive constant A. After choosing a and \u03ba, let A satisfy 2Ae \u2212a > sup x\u2208\u2126 u 0 (x). Then one has",
   "cite_spans":[],"ref_spans":[]},
{
   "section":")S(y, t) dy, 0 < t < T.",
   "text":"Hence S(x, t) is a supersolution to (1.1), and thus inequality (2.3) holds by Theorem 1.1.",
   "cite_spans":[],"ref_spans":[]},
{
   "section":")S(y, t) dy, 0 < t < T.",
   "text":"By Theorem 2.1, we have a supersolution for any T > 0. Hence the local solution u on D T from Theorem 1.2 is bounded in D T for arbitrary T > 0, and thus u can be extended to the whole time domain.",
   "cite_spans":[],"ref_spans":[]},
{
   "section":"Decreasing property of boundary values.",
   "text":"In this section, boundary behavior of the solution to Problem (1.1) is studied. The difference of largest and smallest boundary values grows exponentially (inequality (3.4) ). When the weights are identically zero on some part of boundary, it is shown that the difference can be nonincreasing in Theorem 3.2.",
   "cite_spans":[],"ref_spans":[]}
```

**Comment by dginev** [comment]
Preprocessing fine-print for approaching the arXMLiv 2020 and Stackexchange (kiwix 2020) sets. The usual path has been:
```
HTML -> plaintext -> tfrecords
```

Pretraining an LM is proudly robust to noise and there shouldn't be much need for overthinking the details here. Getting a sane basic setup, that boosts the core textual English + math syntax signal, should be a priority for our purposes. A range of other content can get dropped.

Mostly asking the questions:
 - [ ] Some decisions on how to group non-article genres need to be made. Is an Answer to be embedded in vacuum or with some context e.g. preceded by title? Embedding the entire content of a SE page is not realistic, as most autoencoder models tend to use 512 input tokens.
 - [x] Unicode should be allowed without alterations, at most sanitized for control chars
 - [x] as in 2019, keep in refs and citations, but normalize to a fixed keyword `REF` and `CITE`
 - [x] drop images, tables, footers, headers, any and all metadata, focus on textual content
 - [x] preserve punctuation and letter case
 - [ ] ? include font treatments as we do in math for text? E.g. `𝔄` to `fraktur_U`, and **word** to `bold_word` 
 - [ ] ? include marks for links? I'm definitely dropping the href but say `<a href="wiki...">abelian group</a>` could be normalized to `start_LINK abelian group end_LINK`.
 - [ ] ? include a special word in the vocabulary for paragraph breaks (`\n\n` in plaintext), as an extra cue? Helpful in SE
 - [ ] ? include markers for list items? Maybe a unicode bullet •, to also align with the spurious bullets we see in arXiv?
 - [ ] ? include markers for code blocks (`start_CODE` and `end_CODE` ?)
 - [ ] ? include markers for quotations (`start_BLOCKQUOTE` and `end_BLOCKQUOTE` ?)
 - [ ] ? are we handling footnotes correctly?

Will keep adding here as other bits stream in as I study the data/processing...

**Comment by dginev** [comment]
As I'm looking into the proofwiki data, cataloging some related links with further resources that I stumbled on:

Lists:
1. https://proofwiki.org/wiki/ProofWiki:Community_Portal
2. https://docs.mathjax.org/en/v2.7-latest/misc/mathjax-in-use.html

Resources:

3. https://statproofbook.github.io/I/Table_of_Contents
4. http://www.sklogwiki.org/SklogWiki/index.php/WikiNode
5. https://mathhelpboards.com/forums/pre-algebra-and-algebra.2/
6. http://www.scholarpedia.org/article/Main_Page
7. https://asone.ai/polymath/index.php?title=Main_Page
8. https://oeis.org/wiki/Main_Page
9. https://yutsumura.com/
10. https://www.theoretical-physics.net/dev/index.html
11. https://www.physicsforums.com/
12. https://www.quantumcalculus.org/
13. https://annals.math.princeton.edu/
14. http://www.randomservices.org/random/index.html
15. http://www.fightfinance.com/
16. https://www.intmath.com/

... there's more to expand from here. I won't track them all down now, but it's encouraging to see there is such a wide variety of resources to compile data from, rich in math syntax.
